core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -111,8 +122,11 @@ public BlockInfo getInfo(long blockId) throws IOException {
   /**
    * @return the info of all block workers eligible for reads and writes
    */
-  public List<BlockWorkerInfo> getEligibleWorkers() throws IOException {
-    return getAllWorkers();
+  public synchronized List<BlockWorkerInfo> getEligibleWorkers() throws IOException {
+    if (mWorkerInfoList == null || mWorkerRefreshPolicy.attempt()) {
+      mWorkerInfoList = getAllWorkers();
```

Comment:
would we be able to use this? https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoizeWithExpiration(com.google.common.base.Supplier,%20long,%20java.util.concurrent.TimeUnit)

Reply:
@gpang thanks for the pointer. It seems like handling an exception in the delegated supplier needs some additional logic. I like the code clarity with the current approach so I'll defer to improvement to a later PR